[TIL-265] 관리자 문제 관리(CRUD) 기능 구현

dongyeon-0822 opened the pull request:
## 이슈 번호(링크)

[TIL-265](https://soma-til.atlassian.net/browse/TIL-265)

<br>

## 개요

관리자 문제 관리(CRUD) 기능 구현

<br>

## 내용

1. 문제 생성 기능: `POST /problem` -> [API 문서](https://soma-til.notion.site/24122c52943a4b81a26ab862c6e1ab31?pvs=4)
	• 문제의 제목, 질문, 모범 답안, 난이도, 카테고리 정보를 등록 가능.
	• categoryIdList를 통해 카테고리 유효성 검사 후, 문제와 카테고리 관계 저장.
2. 문제 리스트 조회 기능: `GET /problem` -> [API 문서](https://soma-til.notion.site/20d02c51f5b94b82b6214b2ca7e9123c?pvs=4)
	• 기존에 구현한 페이징, 검색 파라미터 그대로 적용
2. 문제 상세 조회 기능: `GET /problem/{id}` -> [API 문서](https://soma-til.notion.site/ad48a2300b7045b9995184af57465488?pvs=4)
	• 문제 id로 조회 -> 문제의 상세 정보 및 연관된 카테고리 정보를 반환.
	• Problem 엔티티와 연관된 카테고리 정보를 함께 조회하여 반환.
3. 문제 수정 기능: `PUT /problem/{id}` -> [API 문서](https://soma-til.notion.site/1d2024c6d4f748d49c7f6f84142fd5cb?pvs=4)
	• body에 변경 사항이 있는 필드만 전달하면 해당 필드를 업데이트. 
        • 문제 id, 카테고리 id 유효성 검사
	• 문제의 각 필드(title, question, solution, grading, level)에 대해 null 검사를 통해 업데이트할 필드만 적용.
	• categoryIdList가 변경된 경우, 기존 카테고리 관계를 삭제하고 새로운 관계로 갱신.
4. 문제 삭제 기능:  `DELETE /problem/{id}` -> [API 문서](https://soma-til.notion.site/cb5b4a33d0f94b4f9b17e3d5bd625200?pvs=4)
	• 현재는 DB에서 물리적으로 데이터를 삭제 -> soft delete 도입 예정
	• 문제 유효성 검사 후, problemCategory에서 problemid에 해당되는 관계 삭제 후, Problem 삭제.

<br>

## 리뷰어한테 할 말

1. Problem 테이블에 채점 기준을 의미하는 grading이라는 컬럼을 gradingCriteria 로 변경하는게 좋을 것 같습니다! 
    -> grading 이라는 테이블과 이름이 중복되어 헷갈림
2. 문제 생성이나 수정 후에 response에서 생성하거나 수정한 문제 정보를 반환할지 고민이 됬는데, 당장 admin 입장에서 필요하지 않다고 생각하여 제외하고 구현했습니다. 
3. delete는 추후 soft delete로 리팩토링 하면 좋을 것 같습니다. 
4. 문제 id 검증하는 부분은 기존에 구현한 부분과 동일하게 validateProblemExists로 구현한 상태인데, 카테고리 id 검증 같은 경우, problemService에 구현하는 것보다 category repository 단에서 구현하는게 적절하다고 생각해서 따로 분리했습니다. 문제 id 검증도 여러 곳에서 반복해서 사용되기 때문에 일관성 있게 repository 단에 분리하는 건 어떤가요? 


[TIL-265]: https://soma-til.atlassian.net/browse/TIL-265?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

dongyeon-0822 committed:
feat: 관리자 문제 생성(Create) 기능 구현

dongyeon-0822 committed:
feat: 관리자 문제 수정(Update) 기능 구현

dongyeon-0822 committed:
feat: 관리자 문제 삭제(Delete) 기능 구현

dongyeon-0822 committed:
feat: 관리자 문제 리스트, 상세 조회(Read) 기능 구현

CodeDiary18 commented on til-domain/src/main/java/com/til/application/problem/AdminProblemService.java:
```diff
@@ -0,0 +1,84 @@
+                .build();
+
+            problemCategoryRepository.save(problemCategory);
+        }
```
P4: AdminProblemService for문에서 하나하나 데이터를 삽입하지 않고 
ProblemCategoryRepository에 메서드를 생성하여 problemId와 categoryId 리스트를 인자로 넘겨서 저장하는 로직을 구성해보면 좋을 것 같습니다😀

CodeDiary18 commented on til-domain/src/main/java/com/til/domain/category/repository/AdminCategoryRepositoryCustomImpl.java:
```diff
@@ -0,0 +1,30 @@
+        long validCount = queryFactory.select(category.id)
+            .from(category)
+            .where(category.id.in(categoryIdList))
+            .fetchCount();
```
P3: fetchCount()가 deprecated 되어서 아래와 같이 변경해보면 좋을 것 같습니다!
```suggestion
        long validCount = queryFactory.select(category.count())
            .from(category)
            .where(category.id.in(categoryIdList))
            .fetchFirst();
```

CodeDiary18 commented on til-domain/src/main/java/com/til/domain/category/repository/AdminProblemCategoryRepositoryCustomImpl.java:
```diff
@@ -0,0 +1,42 @@
+            ProblemCategory newProblemCategory = ProblemCategory.builder()
+                .problemId(problemId)
+                .categoryId(categoryId)
+                .build();
```
P4: builder가 노출되기 보다는 ProblemCategory 내부에 static 메서드를 정의해서 사용해보면 좋을 것 같습니다!

CodeDiary18 commented on til-domain/src/main/java/com/til/domain/problem/repository/AdminProblemRepositoryCustomImpl.java:
```diff
@@ -0,0 +1,202 @@
+
+        boolean hasUpdate = false;
+
+        if (adminUpdateProblemDto.title() != null ||
+            adminUpdateProblemDto.question() != null ||
+            adminUpdateProblemDto.solution() != null ||
+            adminUpdateProblemDto.grading() != null ||
+            adminUpdateProblemDto.level() != null) {
+
+            if (adminUpdateProblemDto.title() != null) {
+                updateClause.set(problem.title, adminUpdateProblemDto.title());
+                hasUpdate = true;
+            }
+            if (adminUpdateProblemDto.question() != null) {
+                updateClause.set(problem.question, adminUpdateProblemDto.question());
+                hasUpdate = true;
+            }
+            if (adminUpdateProblemDto.solution() != null) {
+                updateClause.set(problem.solution, adminUpdateProblemDto.solution());
+                hasUpdate = true;
+            }
+            if (adminUpdateProblemDto.grading() != null) {
+                updateClause.set(problem.grading, adminUpdateProblemDto.grading());
+                hasUpdate = true;
+            }
+            if (adminUpdateProblemDto.level() != null) {
+                updateClause.set(problem.level, adminUpdateProblemDto.level());
+                hasUpdate = true;
+            }
+        }
```
P4: 제일 처음 if으로 감싸지 않고 내부 if문 만 활용해도 동작이 가능할 것 같습니다!
```suggestion
        if (adminUpdateProblemDto.title() != null) {
            updateClause.set(problem.title, adminUpdateProblemDto.title());
            hasUpdate = true;
        }
        if (adminUpdateProblemDto.question() != null) {
            updateClause.set(problem.question, adminUpdateProblemDto.question());
            hasUpdate = true;
        }
        if (adminUpdateProblemDto.solution() != null) {
            updateClause.set(problem.solution, adminUpdateProblemDto.solution());
            hasUpdate = true;
        }
        if (adminUpdateProblemDto.grading() != null) {
            updateClause.set(problem.grading, adminUpdateProblemDto.grading());
            hasUpdate = true;
        }
        if (adminUpdateProblemDto.level() != null) {
            updateClause.set(problem.level, adminUpdateProblemDto.level());
            hasUpdate = true;
        }
```

CodeDiary18 commented on til-domain/src/main/java/com/til/domain/problem/repository/AdminProblemRepositoryCustomImpl.java:
```diff
@@ -0,0 +1,202 @@
+        return count == null ? 0 : count;
+    }
+
+    private BooleanExpression getSearchCondition(ProblemSearchDto searchDto) {
+        return allOf(
+            getKeywordCondition(searchDto.keyword()),
+            getLevelCondition(searchDto.levelList()),
+            getCategoryCondition(searchDto.categoryList())
+        );
+    }
+
+    private BooleanExpression getKeywordCondition(String keyword) {
+        return hasText(keyword) ? problem.title.containsIgnoreCase(keyword) : null;
+    }
+
+    private static BooleanExpression getLevelCondition(List<Integer> levelList) {
+        return !isNullOrEmpty(levelList) ? problem.level.in(levelList) : null;
+    }
+
+    private BooleanExpression getCategoryCondition(List<Long> categoryList) {
+        return !isNullOrEmpty(categoryList) ? JPAExpressions.selectFrom(problemCategory)
+            .where(problemCategory.problemId.eq(problem.id))
+            .where(problemCategory.categoryId.in(categoryList))
+            .exists() : null;
+    }
```
P4: 조건절 관련해서 클래스를 분리해보면 좋을 것 같습니다!

CodeDiary18 reviewed (commented):
리뷰어한테 하는 말 관련해서 의견을 남깁니다!
> 1. Problem 테이블에 채점 기준을 의미하는 grading이라는 컬럼을 gradingCriteria 로 변경하는게 좋을 것 같습니다!  

-> 모호하다는 의견에 동의해서 변경해보면 좋을 것 같습니다!

> 4. 문제 id 검증하는 부분은 기존에 구현한 부분과 동일하게 validateProblemExists로 구현한 상태인데, 카테고리 id 검증 같은 경우, problemService에 구현하는 것보다 category repository 단에서 구현하는게 적절하다고 생각해서 따로 분리했습니다. 문제 id 검증도 여러 곳에서 반복해서 사용되기 때문에 일관성 있게 repository 단에 분리하는 건 어떤가요?  

-> 어느 수준까지 데이터를 신뢰할 것인지에 대해서도 논의해보고 난 뒤 리팩토링해보면 좋을 것 같습니다!

dongyeon-0822 committed:
refactor: common 모듈 분리에 따른 수정

dongyeon-0822 committed:
refactor: 전반적인 구조 리팩토링

- AdminProblemService for문 -> ProblemCategoryRepository에 저장 메서드 생성
- fetchCount() deprecated 수정 + Null 처리
- builder() -> ProblemCategory에 static 메서드 생성
- 검증 if 문 불필요한 조건 제거
- 리스트 조건 검색 클래스 분리

CodeDiary18 commented on til-admin-api/src/main/java/com/til/controller/problem/ProblemController.java:
```diff
@@ -16,10 +35,47 @@
 @Slf4j
 public class ProblemController {
 
-    @PostMapping("/register")
-    public ApiResponse<Void> createProblem(@CurrentUser Long adminId) {
-        // TODO: 문제 등록 로직 구현
-        log.debug("문제 등록 요청 - 관리자 ID: {}", adminId);
+    private final AdminProblemService adminProblemService;
+
+    @PostMapping("")
+    public ApiResponse<Void> createProblem(@CurrentUser AuthUserInfoDto adminInfo,
```
P1: `@CurrentUser` 사용하는 부분 Long 타입으로 사용자 ID만 받을 수 있어서 아래처럼 변경이 필요합니다!
```java
@CurrentUser Long adminId
```

CodeDiary18 reviewed (commented):
P4: problem 도메인에서 관리자 관련된거는 admin 디렉토리를 생성해 depth를 줘서 구분을 해봐도 좋을 것 같습니다🐧

dongyeon-0822 committed:
refactor: repository 에서 admin 디렉토리 생성

dongyeon-0822 committed:
refactor: @CurrentUser Long 타입으로 변경
